crates/wash-runtime/src/host/http.rs

```diff
@@ -43,11 +41,225 @@ use rustls_pemfile::{certs, private_key};
 use tokio::sync::{RwLock, mpsc};
 use tokio_rustls::TlsAcceptor;
 
-const HTTP_SERVER_ID: &str = "http-server";
+/// Trait defining the routing behavior for HTTP requests
+/// Allows for custom routing logic based on workload IDs and requests
+/// Use this trait to implement custom routing strategies with the default HTTP Extension
+#[async_trait::async_trait]
+pub trait Router: Send + Sync + 'static {
+    /// Register a workload that has been resolved
+    /// and is guaranteed to be available for handling requests
+    async fn on_workload_resolved(
+        &self,
+        resolved_handle: &ResolvedWorkload,
+        component_id: &str,
+    ) -> anyhow::Result<()>;
+
+    /// Unregister a workload that is being stopped
+    async fn on_workload_unbind(&self, workload_id: &str) -> anyhow::Result<()>;
+
+    /// Determine if the outgoing request is allowed
+    fn allow_outgoing_request(
+        &self,
+        workload_id: &str,
+        request: &hyper::Request<wasmtime_wasi_http::body::HyperOutgoingBody>,
+        config: &wasmtime_wasi_http::types::OutgoingRequestConfig,
+    ) -> anyhow::Result<()>;
+
+    /// Pick a workload ID based on the incoming request
+    fn route_incoming_request(
+        &self,
+        req: &hyper::Request<hyper::body::Incoming>,
+    ) -> anyhow::Result<String>;
+}
+
+/// Router that routes requests by 'Host' header, configured via WitInterface config
+#[derive(Default)]
+pub struct DynamicRouter {
+    host_to_workload: tokio::sync::RwLock<HashMap<String, String>>,
+}
+
+/// Implementation of Router that maps Host headers to workload IDs
+/// based on the 'host' config in the wasi:http/incoming-handler interface
+#[async_trait::async_trait]
+impl Router for DynamicRouter {
+    async fn on_workload_resolved(
+        &self,
+        resolved_handle: &ResolvedWorkload,
+        _component_id: &str,
+    ) -> anyhow::Result<()> {
+        let incoming_handler_interface = WitInterface::from("wasi:http/incoming-handler");
+        let Some(http_iface) = resolved_handle
+            .host_interfaces()
+            .iter()
+            .find(|iface| iface.contains(&incoming_handler_interface))
+        else {
+            anyhow::bail!("workload did not request wasi:http/incoming-handler interface");
+        };
+
+        let host_header = http_iface
+            .config
+            .get("host")
+            .cloned()
+            .context("No host header found")?;
+
+        let mut lock = self.host_to_workload.write().await;
+        lock.insert(host_header, resolved_handle.id().to_string());
+
+        Ok(())
+    }
+
+    async fn on_workload_unbind(&self, workload_id: &str) -> anyhow::Result<()> {
+        let mut lock = self.host_to_workload.write().await;
+        lock.remove(workload_id);
```

Comment:
I know this was mostly just moved, but pretty sure this is a bug now that the key can be a host header.

```suggestion
        lock.retain(|_host, wid| wid != workload_id);
```

Alternatively to catch these types of things going forward, we could type these values, e.g. `pub struct HostHeader(String); pub struct WorkloadId(String);`

```rust
host_to_workload: tokio::sync::Mutex<HashMap<HostHeader, WorkloadId>>
```